homotopy-core/src/contraction.rs

```diff
@@ -108,8 +108,9 @@ impl DiagramN {
                 signature,
             )?;
 
-            Ok(vec![cospan])
+            Ok::<_, ContractionError>(vec![cospan])
```

Comment:
Hmm I really don't like that we have to use an `Either` because of this. I think `ContractionError` should contain `DimensionError` but not sure how to change the type of `attach` to only allow `E` types which contain `DimensionError` as a subtype.

Reply:
I'm quite unhappy with the either dependency as well. Can thiserror handle these kinds of things?

Reply:
I have an idea. We can add `DimensionError` to `ContractionError` like this:

```rust
struct ContractionError {
   InvalidBoundaryPath(#[from] DimensionError)
}
```

and that automatically derives `impl From<DimensionError> for ContractionError`. Then we need to change the type of `attach` to add the constraint `E : From<DimensionError>`. That way `attach` can just return `E` instead of an either type.

Reply:
And you won't even need to do any conversions manually because `?` handles all of that for you.

Reply:
One more thing. If you look at `DiagramN::attach`:
https://github.com/homotopy-io/homotopy-rs/blob/ab001c9b512c4f5e352b0d72ded368b856954992/homotopy-core/src/diagram.rs#L450-L469

That already throws a dimension error if the boundary path is invalid. Since attach now handles that, we can remove that check and unify `AttachmentError::Dimension` and `DimensionError`.

Reply:
I did this. I am now happy.